### PR TITLE
fix: allow kernel to read tables with invalid _last_checkpoint

### PR DESCRIPTION
**Pull request by zachschuermann**
Previously, `read_last_checkpoint` would return `Err(<json error>)` if json parsing of the `_last_checkpoint` file failed. This caused the upstream call to fail due to propagation of the error. Instead we now return `None` from `read_last_checkpoint` in this case - similar to what is returned if the `_last_checkpoint` is not found. This allows log replay to continue without the hint and read the table.

Additionally, this unblocks the `corrupted-last-checkpoint-kernel` golden table test.

### CODE REVIEW

File: kernel/src/snapshot.rs [hunk 1]
```diff
@@ -287,7 +287,12 @@ struct CheckpointMetadata {
 
 /// Try reading the `_last_checkpoint` file.
 ///
-/// In case the file is not found, `None` is returned.
+/// Note that we typically want to ignore a missing/invalid `_last_checkpoint` file without failing
+/// the read. Thus, the semantics of this function are to return `None` if the file is not found or
+/// is invalid JSON. Unexpected/unrecoverable errors are returned as `Err` case and are assumed to
+/// cause failure.
+///
+/// TODO: java kernel retries three times before failing, should we do the same?
```
Comment (roeap) [inline kernel/src/snapshot.rs:295]: my initial instinct would be, that if retry made sense, then on io errors when trying to read the file. I would expect this to be the responsibility of of the engine inside `read_files`. This I think is is good as is. OR do we get errors back, that we think are retryable?

**Review by roeap — APPROVED** [review]

File: kernel/src/snapshot.rs (hunk 1 above)
Comment (zachschuermann) [reply kernel/src/snapshot.rs:295]: my initial reaction is also to leave this to the engine. I don't think there is anything specific about reading the _last_checkpoint file and if failures occur across the network it should be handled the same for all our reads and taken care of within `read_files`

File: kernel/src/snapshot.rs [hunk 2]
```diff
@@ -297,7 +302,9 @@ fn read_last_checkpoint(
         .read_files(vec![(file_path, None)])
         .and_then(|mut data| data.next().expect("read_files should return one file"))
     {
-        Ok(data) => Ok(Some(serde_json::from_slice(&data)?)),
+        Ok(data) => Ok(serde_json::from_slice(&data)
+            .map_err(|e| warn!("invalid _last_checkpoint JSON: {e}"))
```
Comment (hntd187) [inline kernel/src/snapshot.rs:306]: You probably want `inspect_err` instead https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect_err

File: kernel/src/snapshot.rs [hunk 3]
```diff
@@ -472,6 +481,47 @@ mod tests {
         assert!(cp.is_none())
     }
 
+    fn valid_last_checkpoint() -> Vec<u8> {
+        r#"{"size":8,"size_in_bytes":21857,"version":1}"#.as_bytes().to_vec()
```
Comment (hntd187) [inline kernel/src/snapshot.rs:485]: you can shorten this to `b#"` instead to make a byte slice and then do `.to_vec()`

File: kernel/src/snapshot.rs [hunk 4]
```diff
@@ -472,6 +481,47 @@ mod tests {
...
+        // in memory file system
+        let store = Arc::new(InMemory::new());
+
+        // put _last_checkpoint file
+        let data = valid_last_checkpoint();
+        let invalid_data = "invalid".as_bytes().to_vec();
+        let path = Path::from("valid/_last_checkpoint");
+        let invalid_path = Path::from("invalid/_last_checkpoint");
+
+        tokio::runtime::Runtime::new()
```
Comment (hntd187) [inline kernel/src/snapshot.rs:499]: can all of this be replace with a `#[tokio::test]` instead?

File: kernel/src/snapshot.rs (hunk 3 above)
Comment (zachschuermann) [reply kernel/src/snapshot.rs:485]: hm don't think I can do `b#`?

File: kernel/src/snapshot.rs (hunk 4 above)
Comment (zachschuermann) [reply kernel/src/snapshot.rs:499]: yea considered doing that but figured this called out async a little better so kept it as-is. I would prefer to have these be sync tests but I think we are forced to async with object_store..

File: kernel/src/snapshot.rs (hunk 3 above)
Comment (zachschuermann) [reply kernel/src/snapshot.rs:485]: decided to leave as-is since doing `b` requirues escaping quotes

File: kernel/src/snapshot.rs (hunk 4 above)
Comment (hntd187) [reply kernel/src/snapshot.rs:499]: Does calling the async out here provide some value to someone trying to understand the test? Assuming they know rust async, doing the runtime plumbing just seems like boiler plate right?

Comment (zachschuermann) [reply kernel/src/snapshot.rs:499]: probably not. I would like to think about how to isolate the async code (really just needed for setup) so this is more of a TODO
